libcasr/src/ubsan.rs

```diff
@@ -0,0 +1,282 @@
+            .collect();
+        // Get description (from first line)
+        let description = message.first().unwrap();
+        let re = Regex::new(r#".+: runtime error: (.+)"#).unwrap();
+        let Some(cap) = re.captures(description) else {
+            return Err(Error::Casr(format!(
+                "Couldn't parse error description: {description}"
+            )));
+        };
+        let description = cap.get(1).unwrap().as_str().to_string();
+        // Get short description (from last line)
+        let short_description = message.last().unwrap();
+        let re = Regex::new(r#"SUMMARY: UndefinedBehaviorSanitizer: (\S+)"#).unwrap();
+        let Some(cap) = re.captures(short_description) else {
+            return Err(Error::Casr(format!(
+                "Couldn't parse ubsan summary: {short_description}"
+            )));
+        };
+        let short_description = cap.get(1).unwrap().as_str().to_string();
+
+        Ok(ExecutionClass::new((
+            "NOT_EXPLOITABLE",
+            &short_description,
+            &description,
+            "",
+        )))
+    }
+}
+
+impl CrashLineExt for UbsanWarning {
+    fn crash_line(&self) -> Result<CrashLine> {
+        let message: Vec<String> = self
+            .message
+            .split('\n')
+            .map(|l| l.trim_end().to_string())
+            .collect();
+        let crashline: String;
+        let mut re = Regex::new(r#".+ (.+):(\d+):\d+"#).unwrap();
+        // If there is no stacktrace use crashline from first string
+        // May be not absolute
+        // Else use first string from stacktrace
+        if message.len() == 2 || message.len() == 3 && message[2].contains(" note: ") {
+            crashline = message[0].clone();
+            re = Regex::new(r#"(.+):(\d+):\d+ runtime error: "#).unwrap();
+        } else if message[1].contains(" note: ") {
+            crashline = message[2].clone();
+        } else {
+            crashline = message[1].clone();
+        }
+
+        // Get file path and offset from crashline
+        let Some(cap) = re.captures(&crashline) else {
+            return Err(Error::Casr(format!(
+                "Couldn't parse error crashline: {crashline}"
+            )));
+        };
+        let file = cap.get(1).unwrap().as_str().to_string();
+        let offset = cap.get(2).unwrap().as_str().parse::<u64>();
+        let Ok(offset) = offset else {
+            return Err(Error::Casr(format!(
+                "Couldn't parse crashline offset: {crashline}"
+            )));
+        };
+        Ok(CrashLine::Module { file, offset })
+    }
+}
+
+/// Extract ubsan warnings form stderr
+///
+/// # Arguments
+///
+/// * `stderr` - output containing ubsan warnings
+///
+/// # Return value
+///
+/// Ubsan warning struct vector
+pub fn extract_ubsan_warnings(stderr: &str) -> Result<Vec<UbsanWarning>> {
+    let mut ubsan_warnings: Vec<UbsanWarning> = vec![];
+    for index in stderr.match_indices("runtime error: ").map(|(i, _)| i) {
```

Comment:
Maybe we can simplify this parsing with regex?

Reply:
No, we can't, I guess, cause UbsanWarning is multiline ubsan message and `*` repetition is greedy in rust regex, but we need to "lazy" parsing